muxers/mplex/src/io.rs

```diff
@@ -0,0 +1,793 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use bytes::Bytes;
+use crate::{MplexConfig, MaxBufferBehaviour};
+use crate::codec::{Codec, Frame, LocalStreamId, RemoteStreamId};
+use log::{debug, trace};
+use fnv::FnvHashMap;
+use futures::{prelude::*, ready, stream::Fuse};
+use futures::task::{ArcWake, waker_ref, WakerRef};
+use futures_codec::Framed;
+use parking_lot::Mutex;
+use std::collections::{VecDeque, hash_map::Entry};
+use std::{cmp, io, mem, sync::Arc, task::{Context, Poll, Waker}};
+
+pub use std::io::{Result, Error, ErrorKind};
+
+/// A multiplexed I/O stream.
+pub struct Multiplexed<C> {
+    /// The current operating status.
+    status: Status,
+    /// The underlying I/O stream.
+    io: Fuse<Framed<C, Codec>>,
+    /// The configuration.
+    config: MplexConfig,
+    /// Buffer of received frames that have not yet been consumed.
+    buffer: Vec<Frame<RemoteStreamId>>,
```

Comment:
In principle, yes, I'd prefer to have a read buffer per stream, i.e. a map. Not just because of potentially suboptimal performance if just a single stream is slow to consume its frames, but also because it would allow just `Reset`ing a single stream if the buffer limit for that stream is hit, instead of failing the entire connection when the shared `buffer` is full with the default configuration `MaxBufferBehaviour::CloseAll`. Essentially changing `MaxBufferBehaviour::CloseAll` to `MaxBufferBehaviour::ResetStream`, as this is more along the lines of what the mplex spec prescibes in the [implementation notes](https://github.com/libp2p/specs/tree/master/mplex#implementation-notes). The configured buffer limit would then apply to a single substream and together with the substream limit lets one still reason about resource usage bounds. But in any case, the shared buffer was pre-existing code and I'd really prefer to propose such a change in a separate PR. I didn't want to pile up too many semantical changes here.